// qgame-toolkit/tools/packager/webpack.config.js
/*
 * Copyright (C) 2017, hapjs.org. All rights reserved.
 */

const path = require('path')
const WebpackZipPlugin = require('./lib/webpack-zip-plugin')
const info = require('./lib/info')
const moduleName = info.moduleName

const FILE_EXT_LIST = info.name.extList

/**
 * 配置关联
 * @param webpackConf
 * @param defaults
 * @param options
 */
function postHook (webpackConf, defaults, options) {
  // 环境信息
  const {
    // 配置环境
    nodeConf,
    appPackageName,
    pathDist,
    pathBuild,
    zipReses
  } = defaults

  webpackConf.module.rules.push(
    {
      test: /\.js/,
      use: [path.resolve(__dirname, 'index.js')],
    },
    {
      test: /\.js/,
      use: ['babel-loader'],
      exclude: function (path) {
        return /node_modules/.test(path) && !(new RegExp(moduleName).test(path))
      }
    }
  )

  // 是否使用release签名
  const dvWithSign = nodeConf.NODE_PHASE === 'dv' && options.sign
  const olWithoutDebug = nodeConf.NODE_PHASE !== 'dv' && !options.debug
  const useReleaseSign = dvWithSign || olWithoutDebug

  webpackConf.plugins.push(
    // 打包
    new WebpackZipPlugin({
      name: appPackageName,
      output: pathDist,
      src: pathBuild,
      res: zipReses,
      sign: useReleaseSign
    })
  )
}

module.exports = {
  postHook
}
